Framework/Core/src/DataSamplingReadoutAdapter.cxx

```diff
@@ -0,0 +1,56 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "Framework/DataSamplingReadoutAdapter.h"
+#include "Headers/DataHeader.h"
+#include "Framework/DataProcessingHeader.h"
+
+namespace o2
+{
+namespace framework
+{
+
+using DataHeader = o2::header::DataHeader;
+
+InjectorFunction dataSamplingReadoutAdapter(OutputSpec const& spec)
+{
+
+  // copied from Common/DataBlock.h
+  using DataBlockId = uint64_t;
+  struct DataBlockHeaderBase {
+    uint32_t blockType;  ///< ID to identify structure type
+    uint32_t headerSize; ///< header size in bytes
+    uint32_t dataSize;   ///< data size following this structure (until next header, if this is not a toplevel block header)
+    DataBlockId id;      ///< id of the block (monotonic increasing sequence)
+    uint32_t linkId;     ///< id of link
+  };
+
+  return [spec](FairMQDevice& device, FairMQParts& parts, int index) {
+    for (size_t i = 0; i < parts.Size() / 2; ++i) {
+
+      auto dbh = reinterpret_cast<DataBlockHeaderBase*>(parts.At(2 * i)->GetData());
```

Comment:
Shouldn't you use `o2::header::get()`.

Reply:
I thought `o2::header::get()` would work only for `header::BaseHeader`children.

Reply:
you are right, of course.

Apart from that, how come this is needed at all? Isn't readout supposed to use the O2 Data Model as well? @mkrzewic